### ISSUE TITLE
LGTM.com - false positive

### ISSUE THREAD

**Issue by mnencia** [github/codeql#4707]
**Description of the false positive**

In verbose regular expressions, spaces are ignored and comments are allowed. The py/regex/unmatchable-dollar and py/regex/unmatchable-caret tests give false positive.

**URL to the alert on the project page on LGTM.com**

https://lgtm.com/projects/g/2ndquadrant-it/barman/snapshot/958a76c31198ee1879d2f89ef418979c034824c7/files/barman/copy_controller.py?sort=name&dir=ASC&mode=heatmap#xfa2c7f56a58e38c9:1

**Comment by RasmusWL** [comment]
Hi @mnencia, thanks for your report :+1: I agree that this is clearly a false positive.

I'm not sure when we'll be able to fix this one, but we will update you on this issue once we have done so :blush:

**Comment by tausbn** [comment]
Just to add a bit more context: the query in question _should_ account for uses of `re.VERBOSE`/`re.MULTILINE`, and so shouldn't flag in this particular case. Off the top of my head, the only reason I can think of for this failing is that we're not (symbolically) evaluating enough of `re.VERBOSE + re.IGNORECASE` to determine that `re.VERBOSE` indeed holds for this regex. (Or possibly it's because these flags are combined using `+` rather than `|`. I'm pretty sure we account for the latter, but I can't immediately say that we're accounting for the former as well.)

**Comment by RasmusWL** [comment]
Hi @mnencia, I had time to look more into this now. The [official documentation](https://docs.python.org/3.9/library/re.html#re.compile) says:

> The expression’s behaviour can be modified by specifying a flags value. Values can be any of the following variables, combined using bitwise OR (the | operator).

So you can get this alert to go away by using `|` instead :wink: 

While it would be nice for our analysis to also handle `+`, which does indeed work in your situation, it's not going to be a high priority to fix this.

**Comment by cosinekitty** [comment]
I have also hit this case, without using the `+` operator to combine flags. See here:

https://lgtm.com/projects/g/cosinekitty/astronomy/snapshot/7687aa49751f8d11129b6fb01347a83d1e585e2a/files/generate/eclipse/norm.py?sort=name&dir=ASC&mode=heatmap#xd3352262c13566d0:1

I believe it is picking up the `[11]` toward the end of my regex text, thinking that it is a character class, when instead it is a comment. I could easily work around this by reformatting my comments, but I thought it might be helpful to report this to the LGTM team. I formatted these numbers in square brackets as a cue to how to use the resulting compiled regex to extract the groups it captures.

**Comment by tausbn** [comment]
Thank you for your report!

This is a slightly different issue (though again related to how we handle `re.VERBOSE`). In this case, the issue stems from the fact that when we encounter the `re.VERBOSE` flag for this query, we only ignore multiple occurrences of [whitespace](https://github.com/github/codeql/blob/main/python/ql/src/Expressions/Regex/DuplicateCharacterInSet.ql#L35) inside a character class.
Ideally, we should be ignoring everything that appears inside a comment as well.

We're currently focused on improving our security queries, and as this isn't immediately related to security, I can't give an estimate for when this will be fixed. We will be tracking this issue and fixing it eventually, but in the mean time my best suggestion would be to suppress the alert (or the query in its entirety) as described in the LGTM.com help.

**Comment by cosinekitty** [comment]
OK, that's fine. I will just ignore it. My wife says I'm good at that.

**Comment by tausbn** [comment]
Yeah, I think that's spot on. We do account for _some_ characters (i.e. whitespace) in regexes with the `re.VERBOSE` flag, but we have no handling of comments. 

Thank you for the report! We'll look into whether there's an easy way to address this.
